### ISSUE TITLE
Running into a NaN problem

### ISSUE THREAD

**Issue by pierregodard** [neulab/xnmt#399]
I apologize if this belongs to dynet but I am running into a NaN problem after a small modification of the attender. I am trying to bias the attention towards longer source words. My config file is similar to `01_standard.yaml`, with added reporting and forced decoding options. The training seems to start normally but after a certain time (sometimes after more than 100 epochs) I get the following error:
```
File "[...]/xnmt/lstm.py", line 163, in add_input_to_prev
self.weightnoise_std if self.train else 0.0)
File "_dynet.pyx", line 4406, in _dynet.vanilla_lstm_gates_dropout_concat
File "_dynet.pyx", line 4439, in _dynet.vanilla_lstm_gates_dropout_concat
RuntimeError: NaN or Inf detected
```
(coming from `MlpSoftmaxDecoderState(...)`)

I have read the discussions [here](https://github.com/clab/dynet/issues/711) and [here](https://github.com/clab/dynet/issues/990) but I can't see how I could end up out of the domain of a function inside `dynet.vanilla_lstm_gates_dropout_concat` as this function computes mostly several sigmoids and a tanh if I am not mistaken.

Mostly my modification of `calc_context` in `attender.py` consists in replacing
```python
return I * attention
```
with
```python
lengths = np.array([float(len(w)) for w in self.curr_words])  # I set self.curr_words in init_sent
norm_avg_lengths = lengths / np.mean(lengths)
v = norm_avg_lengths.reshape(len(self.curr_words), 1)  # This would be a problem with a batch size different than 1 but I am not batching yet
L = dy.inputTensor(v)
weighted_attention = dy.cmult(attention, L)
return I * weighted_attention
```
There might be something silly in what I'm doing and I would appreciate it very much if someone could take a quick look and give me some advice to debug (I am already using immediate computation and validity check for dynet with the `--settings=debug` command line argument in `xnmt`). 

Thank you for your help.

**Comment by pierregodard** [comment]
Thank you very much for taking a look and for this insight @msperber. I confirm that I'm observing this on CPU. I will try to reproduce it on GPU, although I won't run my current experiments on GPU (I'm working on a very small quantity of data, and training on CPU is in the hour range and provides me much more flexible access to computational resources).

**Comment by neubig** [comment]
We should really fix this though, so if we can find a way to reproduce this that'd be good

**Comment by pierregodard** [comment]
Reproducing seems easy on my end, each run of more than 50 epochs eventually produces that error on CPU. Would that be helpful if I moved my data, code branch and config file to the Bridges servers we used during JSALT to try to reproduce this on CPU there? -- in which case taking a look could be easier for you. (In any case I might be doing this anyway to follow Matthias' advice to try this on GPU.) 
Thank you for your help!
